File: src/std/js/NioBufPeer.js


fan.std.NioBufPeer.prototype.init = function(self, file, mode, pos, size) {
	var reader = new FileReader();
	var buf = reader.readAsArrayBuffer(file);
	this.m_fp = new DataView(buf, pos, pos + size);
	this.m_pos = 0;
}

fan.std.NioBufPeer.prototype.size() = function(self) {
	this.fp.byteLength;
}

fan.std.NioBufPeer.prototype.size$() = function(self, x) {
}

fan.std.NioBufPeer.prototype.capacity() = function(self) {
	throw MAX_VALUE;
}

fan.std.NioBufPeer.prototype.capacity$() = function(self, capa) {
}

fan.std.NioBufPeer.prototype.pos() = function(self) {
	return this.m_pos;
}

fan.std.NioBufPeer.prototype.pos$() = function(self, x) {
	this.m_pos = x;
}

fan.std.NioBufPeer.prototype.getByte() = function(self, pos) {
	return this.m_pf.getUint8(pos)
}

fan.std.NioBufPeer.prototype.setByte() = function(self, pos, b) {
	this.m_pf.setUint8(pos, b);
}

fan.std.NioBufPeer.prototype.getBytes() = function(self, pos, dst, off, len) {
	var size = this.m_pf.byteLength-pos;
	if (size > len) size = len;

	for (var i=0; i<size; ++i) {
		dst.set(off+i, this.m_pf.getUint8(pos+i));
	}
	return size;
}

fan.std.NioBufPeer.prototype.setBytes() = function(self, pos, src, off, len) {
	var size = this.m_pf.byteLength-pos;
	if (size > len) size = len;

	for (var i=0; i<size; ++i) {
		this.m_pf.setUint8(pos+i, dst.get(off+i));
	}
	return size;
}

fan.std.NioBufPeer.prototype.close() = function(self) {
}

fan.std.NioBufPeer.prototype.sync() = function(self) {
}

File: library/std/java/src/fan/std/NioBufPeer.java
package fan.std;

import java.io.FileNotFoundException;
import java.io.RandomAccessFile;
import java.nio.ByteBuffer;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.channels.FileChannel.MapMode;

import fan.sys.ArgErr;
import fan.sys.IOErr;
import fan.sys.UnsupportedErr;

public class NioBufPeer {
	ByteBuffer buf;

	public static NioBufPeer make(NioBuf buf) {
		return new NioBufPeer();
	}
	
	public static NioBuf fromJava(java.nio.ByteBuffer buf) {
		NioBuf fbuf = NioBuf.make();
		((NioBufPeer)fbuf.peer).buf = buf;
		return fbuf;
	}

	void alloc(NioBuf self, long size) {
		buf = ByteBuffer.allocate((int)size);
	}

	void init(NioBuf self, File file, String mode, long pos, Long size) throws FileNotFoundException {
		LocalFile lfile = (LocalFile) file;
		try {
			// map mode
			String rw;
			MapMode mm;
			if (mode.equals("r")) {
				rw = "r";
				mm = MapMode.READ_ONLY;
			} else if (mode.equals("rw")) {
				rw = "rw";
				mm = MapMode.READ_WRITE;
			} else if (mode.equals("p")) {
				rw = "rw";
				mm = MapMode.PRIVATE;
			} else
				throw ArgErr.make("Invalid mode: " + mode);

			// if size is null, use file size
			if (size == null)
				size = file.size();

			// traverse the various Java APIs
			RandomAccessFile fp = null;
			FileChannel chan = null;
			try {
				fp = new RandomAccessFile((java.io.File) lfile.jfile, rw);
				chan = fp.getChannel();
				buf = (chan.map(mm, pos, size.longValue()));
			} finally {
				if (chan != null)
					chan.close();
				if (fp != null)
					fp.close();
			}
		} catch (java.io.IOException e) {
			throw IOErr.make(e);
		}
	}

	long size(NioBuf self) {
		return buf.limit();
	}

	public final void size(NioBuf self, long x) {
		buf.limit((int) x);
	}

	long capacity(NioBuf self) {
		return buf.capacity();
	}

	void capacity(NioBuf self, long capa) {
		throw UnsupportedErr.make("mmap capacity fixed");
	}

	long pos(NioBuf self) {
		return buf.position();
	}

	void pos(NioBuf self, long x) {
		buf.position((int) x);
	}

	long getByte(NioBuf self, long pos) {
		return buf.get((int) pos) & 0xff;
	}

	void setByte(NioBuf self, long pos, long b) {
		buf.put((int) pos, (byte) b);
	}

	long getBytes(NioBuf self, long pos, byte[] dst, long off, long len) {
		int oldPos = buf.position();
		buf.position((int) pos);
		buf.get(dst, (int) off, (int) len);
		buf.position(oldPos);
		return len;
	}

	void setBytes(NioBuf self, long pos, byte[] src, long off, long len) {
		int oldPos = buf.position();
		buf.position((int) pos);
		buf.put(src, (int) off, (int) len);
		buf.position(oldPos);
	}

	boolean close(NioBuf self) {
		return true;
	}

	Buf sync(NioBuf self) {
		if (buf instanceof MappedByteBuffer)
			((MappedByteBuffer) buf).force();
		return self;
	}

	public ByteBuffer toByteBuffer() {
		//return buf.duplicate();
		return buf;
	}
}
